Source/Dafny/Resolver.cs

```diff
@@ -5032,6 +5030,26 @@ public class XConstraint {
         bool satisfied;
         Type tUp, uUp;
         switch (ConstraintName) {
+          case "SubsetTypeOfCompilable": {
+              var u = Types[0].NormalizeExpandKeepConstraints();
+              if (!CheckTypeInference_Visitor.IsDetermined(u)) {
+                if (Types[1] is TypeProxy tp && tp.IsTypeParameter) {
+                  resolver.ConstrainSubtypeRelation_Equal(u, tp, errorMsg);
+                  convertedIntoOtherTypeConstraints = true;
+                  return true;
+                }
+                return false;
+              }
+
+              t = Types[1].NormalizeExpandKeepConstraints();
+              if (CheckTypeInference_Visitor.IsDetermined(t)) {
+                // No need to add other constraints, the verifier will take care of that.
+                return true;
+              }
+              resolver.ConstrainSubtypeRelation(t, u.GetCompilableParentType(), errorMsg, true);
```

Comment:
Linking back to where the constraint SubsetTypeOfCompilable is defined, we have:

Types[0] = u = v.Type = requestedType
Types[1] = t = collectionVarType = collectionElementType

So, the added constraint states that the collectionVarType should be a subtype of the compiled supertype of the requested type, which is _exactly_ what you expected.
With the renaming I'm pushing, it should be clearer.

Reply:
Yikes! You're right, the supertype arguably comes first based on the signature, which I did not expect by reading the code.
It probably did not bother in practice because there was only one type for the collection type and it chose it (whether it was the upper or lower bound), but I see how it could create errors. Let me fix that.